### PR TITLE
Global mutex with atomics

### PR DESCRIPTION
**Pull request by alex-virodov**
This is a solution to #16. See comments in `libisyntax_get_global_mutex()` for further implementation details.

### CODE REVIEW

**Comment by alex-virodov** [comment]
@bgilbert fyi and in case you want to comment as well, your comments are very welcomed.

File: test/thread_test.c [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    // Spawn the threads.
+    for (int thread_i = 0; thread_i < n_threads; ++thread_i) {
+
+      threads[thread_i].atomic_sync_flag = &atomic_sync_flag;
+      threads[thread_i].func = func;
+      threads[thread_i].arg = arg;
+      threads[thread_i].force_sync = force_sync;
+
+      assert(thrd_create(&threads[thread_i].thread_id, parallel_sync_and_call, &threads[thread_i]) == thrd_success);
```
Comment (Falcury) [inline test/thread_test.c]: Will get optimized out in Release builds.

**Comment by Falcury** [comment]
So, do I understand correctly you would prefer continuing to use the existing benaphore implementation, instead of replacing it with something else, wrapping SRWLock etc.?

So far I've been using an atomic_compare_exchange in intrinsics.h, that doesn't use stdatomic.h.
However, I suppose it might also be fine to do it this way. I'm not too familiar (yet) with stdatomic.h / threads.h. Has the C11/C17 interface matured enough these days to use those features safely? (I think directly calling into pthread and Win32 API used to be the 'preferred' way?)

**Comment by alex-virodov** [comment]
> Has the C11/C17 interface matured enough these days to use those features safely? 

Unfortunately not, I just haven't touched windows development for a while, so I forgot how MSVC doesn't support standards.
FYI stdatomics.h is in preview and threads.h is not even there: https://devblogs.microsoft.com/cppblog/c11-atomics-in-visual-studio-2022-version-17-5-preview-2/

I will reimplement using intrinsics.h and will test on both windows and linux before I proceed. I may need to refactor thread support from libisyntax.c just so we have a uniform way of managing threads.

**Comment by alex-virodov** [comment]
> So, do I understand correctly you would prefer continuing to use the existing benaphore implementation, instead of replacing it with something else, wrapping SRWLock etc.?

Yes, I like your idea of using atomics for initial sync (I was thinking the same), so as long as you are ok with the general implementation (specifically spinlock is the suspicious part), I'd like to proceed without adding any new primitives to the codebase.

**Comment by Falcury** [comment]
> Yes, I like your idea of using atomics for initial sync (I was thinking the same), so as long as you are ok with the general implementation (specifically spinlock is the suspicious part), I'd like to proceed without adding any new primitives to the codebase.

OK! Yes, the spinlock seems fine to me, I don't suppose we have a good way around it and it should be rare (and brief) to end up there.

For Slidescape I have been using the `work_queue_t` for both distributing jobs and collecting the result of finished jobs (with options for callbacks, manual querying, or both). I have been refactoring it a bit to enable a test program for `isyntax_streamer.c`. Still working on the test program but I already pushed a commit for the work queue refactoring (should hopefully make it a bit easier to work with, if you want to use it).

**Comment by alex-virodov** [comment]
> but I already pushed a commit for the work queue refactoring 

Problem is that I can't use work_queue to test work_queue init :)
I was thinking moving the thread creation code in init_thread_pool() into platform.c/.h or other appropriate place so I can reuse in tests.
Alternative is to test with c++ or test only on some platform (e.g. linux that has <thread.h>).

**Comment by Falcury** [comment]
> Problem is that I can't use work_queue to test work_queue init :) I was thinking moving the thread creation code in init_thread_pool() into platform.c/.h or other appropriate place so I can reuse in tests. Alternative is to test with c++ or test only on some platform (e.g. linux that has <thread.h>).

Sure, no problem to move init_thread_pool() to another location. Could be platform.c, or alternatively maybe a new file platform/thread_pool.c/.h? Do you want me to do the refactor on the main branch, or do you want to do it?

File: test/thread_test.c (hunk 1 above)
Comment (alex-virodov) [reply test/thread_test.c]: Fixed, thank you for catching this.

**Comment by alex-virodov** [comment]
> Sure, no problem to move init_thread_pool() to another location. Could be platform.c, or alternatively maybe a new file platform/thread_pool.c/.h? Do you want me to do the refactor on the main branch, or do you want to do it?

I'd suggest to keep it simple - you keep doing refactor of streamer as you see fit, and no need to move the init code if it doesn't help you. 

For now I disabled the test on windows. I'll come back to that when I do threads within `libisyntax_tile_read()`. Maybe by that time the streamer refactoring will be complete and we won't have many things in flight in parallel. Or please let me know if you prefer something else.

This PR is ready for another review, when convenient. Thank you.

**Review by Falcury — APPROVED** [review]

**Pull request merged**